Add register_table procedure support for iceberg using existing metadata

krvikash opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Adding `register_table` procedure for iceberg connector to register table using existing metadata.

- Iceberg table will be created using provided table location
- Look for the latest metadata file ending with `.metadata.json` inside the `metadata` folder and use it for creating the table
- If provided location is invalid/does not exist then throw an exception
- If no metadata file exists inside provided table location then throw an exception
- If more than one metadata file exists with the latest sequence number then throw an exception
- `schema_name`, `table_name`, and `table_location` should be not-null and valid, Otherwise, the exception will be thrown
-  User can optionally provide the valid `metadata_file_location` (See the usage below)

Valid usages:
 - `register_table(schema_name => ..., table_name => ..., table_location => ...)`
 - `register_table(schema_name => ..., table_name => ..., table_location => ..., metadata_file_location => ...)`

 Sample Queries:
`CALL iceberg.system.register_table('default', 'src_22', 'hdfs://hadoop-master:9000/user/hive/warehouse/orders_5-581fad8517934af6be1857a903559d44');`
 
`CALL iceberg.system.register_table('default', 'src_22', 'hdfs://hadoop-master:9000/user/hive/warehouse/orders_5-581fad8517934af6be1857a903559d44', null);`

 `CALL iceberg.system.register_table('default', 'src_22', 'hdfs://hadoop-master:9000/user/hive/warehouse/orders_5-581fad8517934af6be1857a903559d44', '00003-409702ba-4735-4645-8f14-09537cc0b2c8.metadata.json');`

  TODO:
  Add test cases

<!-- Provide a user-friendly explanation, keep it brief if it isn't user-visible. -->
## Non-technical explanation

NA

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(X) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

findepi commented:
Added empty commit to retrigger the build after https://github.com/trinodb/trino/pull/14323 merged.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:
```diff
@@ -623,4 +636,111 @@ private static void validateOrcBloomFilterColumns(ConnectorTableMetadata tableMe
+    {
+        try {
+            if (!fileSystem.newInputFile(location).exists()) {
+                throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, format("Location %s does not exist", location));
```
The error code is not appropriate. The file  system is fine, is just that the file at the location provided is missing.

findinpath commented:
Let's add some tests.

Note that `io.trino.plugin.iceberg.catalog.TrinoCatalog#dropTable` implementation do delete the table data (and metadata) as well.

The metadata files contain full path names for the `manifest-list` property.
One way to test this functionality is to create a table via SQL and **programatically** drop its content from the metastore so that we avoid to delete the content of the table . In this fashion, the table data remains intact and we can proceed to register it again and eventually make queries on the table.

You can find inspiration on how to get access within the integration tests to the metastore / aws client directly by looking at these tests:

- https://github.com/trinodb/trino/blob/5ac5fdb5f079f532b28128bf0bf9bf158fb8d55f/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/file/TestIcebergFileMetastoreTableOperationsInsertFailure.java
- https://github.com/trinodb/trino/blob/7893908c1afe4da84faa245494f3ee4c6e458ccd/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueTableOperationsInsertFailure.java

krvikash commented:
This PR got closed because of my mistake during running GitHub commands. Reopen a new PR https://github.com/trinodb/trino/pull/14325

@findinpath, Addressed your review comments in https://github.com/trinodb/trino/pull/14325

alexjo2144 commented:
> This PR got closed because of my mistake during running GitHub commands

For next time, you can reopen the same PR. There's a button at the bottom of the page
